File: client/src/js/print-relay.js
import { API_BASE_URL, session } from "./api.js";

const KEY = "educa-pos-print-relay-session";
const id = () => localStorage.getItem(KEY) || (localStorage.setItem(KEY, crypto.randomUUID()), localStorage.getItem(KEY));

export function createPrintRelay({ device, storeId, printerStatus, onTargets, onRequest, onResult }) {
  let socket, ref = 1, joinRef = null, presence = {}, stopped = false;
  const callbacks = new Map(), topic = `print-relay:${storeId}`;
  const targets = () => Object.entries(presence).flatMap(([session_id, value]) => (value.metas || []).filter((m) => m.device === "desktop" && m.printer_online).map((m) => ({ session_id, ...m })));
  const notify = () => onTargets(targets());
  const send = (event, payload, cb) => {
    const messageRef = String(ref++);
    if (event === "phx_join") joinRef = messageRef;
    if (cb) callbacks.set(messageRef, cb);
    socket?.send(JSON.stringify([joinRef || messageRef, messageRef, topic, event, payload]));
  };
  const connect = async () => {
    if (stopped || !session()?.token) return;
    const url = new URL(API_BASE_URL); url.protocol = url.protocol === "https:" ? "wss:" : "ws:"; url.pathname = "/socket/websocket"; url.search = `vsn=2.0.0&token=${encodeURIComponent(session().token)}`;
    socket = new WebSocket(url);
    socket.onopen = async () => { const status = device === "desktop" ? await printerStatus().catch(() => ({ connected: false })) : {}; send("phx_join", { device, session_id: id(), printer_online: !!status.connected, printer: status.model, label: device === "desktop" ? "Desktop Tauri" : "Mobile POS" }); };
    socket.onmessage = ({ data }) => { const [, messageRef, , event, payload] = JSON.parse(data); if (event === "phx_reply") { const cb = callbacks.get(messageRef); callbacks.delete(messageRef); cb?.(payload); if (payload.response?.targets) onTargets(payload.response.targets); } else if (event === "presence_state") { presence = payload; notify(); } else if (event === "presence_diff") { Object.assign(presence, payload.joins); Object.entries(payload.leaves || {}).forEach(([key, value]) => { const refs = new Set(value.metas.map((m) => m.phx_ref)); const metas = (presence[key]?.metas || []).filter((m) => !refs.has(m.phx_ref)); metas.length ? presence[key] = { metas } : delete presence[key]; }); notify(); } else if (event === "print_request") onRequest?.(payload); else if (event === "print_result") onResult?.(payload); };
    socket.onclose = () => { presence = {}; notify(); if (!stopped) setTimeout(connect, 2_000); };
  };
  return { connect, close: () => { stopped = true; socket?.close(); }, requestPrint: (request_id, target_session_id, receipt) => new Promise((resolve, reject) => send("print", { request_id, target_session_id, receipt }, (reply) => reply.status === "ok" ? resolve(reply.response) : reject(new Error(reply.response?.reason || "Print request failed.")))), reportResult: (payload) => send("print_result", payload), updatePrinter: async () => { if (device !== "desktop" || socket?.readyState !== WebSocket.OPEN) return; const s = await printerStatus().catch(() => ({ connected: false })); send("printer_status", { printer_online: !!s.connected, printer: s.model, label: "Desktop Tauri" }); } };
}

File: server/lib/pos_server/print_relay.ex
defmodule PosServer.PrintRelay do
  @moduledoc "Tenant-scoped topics and availability lookups for mobile receipt printing."

  alias PosServerWeb.Presence

  def topic(tenant, store_id), do: "print-relay:#{tenant}:#{store_id}"
  def desktop_topic(tenant, store_id, session_id), do: "#{topic(tenant, store_id)}:desktop:#{session_id}"
  def result_topic(tenant, store_id, request_id), do: "#{topic(tenant, store_id)}:result:#{request_id}"

  def available_desktops(tenant, store_id) do
    tenant
    |> topic(store_id)
    |> Presence.list()
    |> Enum.flat_map(fn {session_id, %{metas: metas}} ->
      metas
      |> Enum.filter(&(&1.device == "desktop" and &1.printer_online == true))
      |> Enum.map(fn meta ->
        %{
          session_id: session_id,
          label: meta.label || "Desktop",
          printer: meta.printer || "Receipt printer",
          user: meta.user,
          online: true
        }
      end)
    end)
  end

  def available_desktop?(tenant, store_id, session_id) do
    Enum.any?(available_desktops(tenant, store_id), &(&1.session_id == session_id))
  end
end
